provider/pkg/resources/resources.go

```diff
@@ -28,15 +26,75 @@ type CloudAPIMetadata struct {
 	Functions map[string]CloudAPIFunction `json:"functions"`
 }
 
-// CloudAPIEndpoint is an endpoint template and the values needed to instantiate the template.
+// CloudAPIEndpoint contains information needed to construct a REST endpoint.
 type CloudAPIEndpoint struct {
+	// SelfLinkProperty is the name of the resource property containing a selfLink.
+	// Example: `selfLink`
+	SelfLinkProperty string `json:"selfLinkProperty,omitempty"`
 	// Template is the template string for a REST endpoint.
-	// Example: v1/projects/{projectsId}/locations/{locationsId}/registries/{registriesId}
+	// Example: https://sqladmin.googleapis.com/v1/projects/{project}/instances/{instance}/sslCerts
 	Template string `json:"template,omitempty"`
 	// Values are the concrete values needed to instantiate the template.
 	Values []CloudAPIResourceParam `json:"values,omitempty"`
 }
 
+// URI constructs a concrete URI value based on the properties of a created resource.
+func (e CloudAPIEndpoint) URI(
+	inputs map[string]interface{},
+	outputs map[string]interface{},
+) (string, error) {
+	if len(e.SelfLinkProperty) > 0 {
+		v, ok := outputs[e.SelfLinkProperty].(string)
+		if !ok {
+			return "", fmt.Errorf("selfLink property %q not found", e.SelfLinkProperty)
+		}
+		return v, nil
+	}
+
+	id := e.Template
+	idParams := e.Values
+	for _, param := range idParams {
+		var propValue string
+		if v, has := evalPropertyValue(inputs, param.SdkName); has {
+			propValue = v
+		} else if v, has := evalPropertyValue(outputs, param.SdkName); has {
```

Comment:
Just trying to understand why/when we would want to fall back to the old value? Is that the actual desired behavior we want?

Reply:
I copied this logic from what we had before. I'm not entirely sure. Perhaps it has to do with properties like `selfLink` that are populated after creation?

Reply:
Confirmed and added a test